### ISSUE TITLE
Inconsitency between number of frames by ffmpeg and VideoIO

### ISSUE THREAD

**Issue by yakir12** [JuliaIO/VideoIO.jl#209]
Counting the number of frames with `ffmpeg` matches the number of images that are in the video, `VideoIO` misses the mark by 2 frames.  

MWE:
```julia
# here we create `n` frames in the form of individual images
using FileIO, Colors, Printf
n = 10
path = mktempdir()
for i in 1:n
    name = @sprintf "%03i.png" i
    img = rand(RGB, 10, 10)
    FileIO.save(joinpath(path, name), img)
end

# here we convert these `n` images into a video
x = joinpath(path, "%03d.png")
y = joinpath(path, "out.mp4")
run(`ffmpeg -i $x -pix_fmt yuv420p $y`)

# now let's see how many frames VideoIO thinks are in this video
using VideoIO
f = VideoIO.openvideo(y)
fn = 0
while !eof(f)
    global fn
    read(f)
    fn += 1
end
@show fn # I get `n - 2` here

# and how many ffmpeg thinks
run(`ffmpeg -i $y -map 0:v:0 -c copy -f null -`) # I get `n`
```

**Comment by ianshmean** [comment]
Interesting. Could you modify the (great) MWE such that each frame has a unique pixel value, and then inspect the values VideoIO is reading in, to see which frames are missing? I.e. two off the end or start, or one off each?

**Comment by yakir12** [comment]
I have to run, but something more than just VideoIO is weird with this:
```
# here we create `n` frames in the form of individual images
using FileIO, Colors, Printf
n = 11
levels = range(0, stop = 1, length = n)
path = mktempdir()
for (i, l) in enumerate(levels)
    name = @sprintf "%03i.png" i
    img = fill(Gray(l), 100, 100)
    FileIO.save(joinpath(path, name), img)
end

# here we convert these `n` images into a video
x = joinpath(path, "%03d.png")
y = joinpath(path, "out.mp4")
run(`ffmpeg -i $x -pix_fmt gray $y`)

# now let's see how many frames VideoIO thinks are in this video
using VideoIO, Statistics
f = VideoIO.openvideo(y)
fn = 0
pix = Float64[]
while !eof(f)
    global fn
    img = read(f)
    push!(pix, mean(red.(img)))
    fn += 1
end
@show fn # I get `n - 2` here
@show pix # crazy

# and how many ffmpeg thinks
run(`ffmpeg -i $y -map 0:v:0 -c copy -f null -`) # I get `n`
```

**Comment by yakir12** [comment]
I have some issues with saving `.png` files, but with `.jpg` this works well. 

In the following MWE you'll see that `VideoIO` ignores the **last two frames** (so not 2 in the beginning, or one off the start and one of the end):

```julia
# here we create `n` frames in the form of individual images
using FileIO, Colors, Printf
n = 6
levels = UInt8.(range(0, stop = 255, length = n))
path = mktempdir()
img = Matrix{UInt8}(undef, 10, 10)
for (i, l) in enumerate(levels)
    name = @sprintf "%03i.jpg" i
    fill!(img, l)
    save(joinpath(path, name), img)
end

# here we convert these `n` images into a video
x = joinpath(path, "%03d.jpg")
y = joinpath(path, "out.mp4")
run(`ffmpeg -i $x -pix_fmt gray $y`)

# now let's see how many frames VideoIO thinks are in this video
# and what images "made it"
using VideoIO, Statistics
f = VideoIO.openvideo(y)
fn = 0
pix = Float64[]
while !eof(f)
    global fn
    img = read(f)
    push!(pix, mean(Gray.(img)))
    fn += 1
end
@show fn # I get `n - 2` here
@show pix # I get [0.0, 0.2, 0.4, 0.6] So the last 2 are missing
```

**Comment by yakir12** [comment]
La bumpa.

**Comment by yakir12** [comment]
Just wanted to add, if I reduce the intended number of frames, `n`, all the way to 1, `ffmpeg` keeps up, while `VideoIO` insists on `n-2`. When `1 ≤ n  ≤ 2` then `VideoIO` claims there are 0 frames in that object (`eof`returns true for the first read). 

The fact it's the last (two) frames, and not the first seems strange. I'd imagine that somewhere during `VideoIO.openvideo` two frames are used for some metadata and thus dropped. But these are the **last** two frames, so no idea how this would happen.

**Comment by ianshmean** [comment]
Have you tried encoding the video with videoio? The package tests do a load of video encoding and checking for frame accuracy, so what you’re seeing doesn’t happen in package tests.

**Comment by yakir12** [comment]
I did not, until now. And the results are even "worse", now it looks like VideoIO is dropping 4 frames...

```julia
using FileIO, VideoIO, Colors, Printf, Statistics

# create the image stack, with `n` frames
n = 9
imgstack = [fill(round(UInt8, l), 10, 10) for l in range(0, stop = 255, length = n)]

# encode the frames into a video using VideoIO
path = mktempdir()
props = [:priv_data => ("crf"=>"22","preset"=>"medium")]
y = joinpath(path, "out.mp4")
encodevideo(y,imgstack,framerate=30,AVCodecContextProperties=props)

# now let's see how many frames VideoIO thinks are in this video
# and what images "made it"
f = VideoIO.openvideo(y)
fn = 0
pix = Float64[]
while !eof(f)
    global fn
    img = read(f)
    push!(pix, mean(Gray.(img)))
    fn += 1
end
@show fn # I get `n - 4` here
@show pix # I get really different things here now

# and how many ffmpeg thinks
run(`ffmpeg -i $y -map 0:v:0 -c copy -f null -`) # I get `n`
```

Gonna take a look at the tests...

**Comment by yakir12** [comment]
Thanks to @ianshmean I got this issue resolved. It was all in the encoding settings:

```julia
using VideoIO

# create the image stack, with `n` frames
n = 6
levels = round.(range(0, stop = 255, length = n))
imgstack = [fill(UInt8(l), 10, 10) for l in levels]

# encode the frames into a video using VideoIO
path = mktempdir()
props = [:color_range=>2, :priv_data => ("crf"=>"0","preset"=>"medium")]
y = joinpath(path, "out.mp4")
encodevideo(y,imgstack,framerate=30,AVCodecContextProperties=props)

# now let's see how many frames VideoIO thinks are in this video
# and what images "made it"
f = VideoIO.openvideo(y, target_format=VideoIO.AV_PIX_FMT_GRAY8)
fn = 0
pix = Float64[]
while !eof(f)
    global fn
    img = reinterpret(UInt8, read(f))
    push!(pix, img[1])
    fn += 1
end
@assert fn == n
@assert pix == levels
```

I'll therefore close this.

**Comment by yha** [comment]
Should this really be closed, just because the issue does not occur for some specific encoding options? It still seems to be skipping frames in other cases. 
I found that videos encoded from MATLAB with the default settings consistently have one frame missing when decoded with `VideoIO`.

**Comment by ianshmean** [comment]
Yeah we should reopen. 

We should be looking for examples of this with the ffmpeg c api in the wild, like this: https://stackoverflow.com/questions/45177897/ffmpeg-libx264-c-api-frames-dropped-from-end-of-short-mp4

**Comment by yha** [comment]
> We should be looking for examples of this with the ffmpeg c api in the wild, like this: https://stackoverflow.com/questions/45177897/ffmpeg-libx264-c-api-frames-dropped-from-end-of-short-mp4

It looks like this is about frames not being encoded into the video, whereas this issue is about frames that are encoded into the video but not returned by the decoder.

**Comment by IanButterworth** [comment]
Should be fixed in v0.9

**Comment by yakir12** [comment]
I can confirm that the first MWE in this issue is resolved with `v0.9.1`.
